examples/screencopy-dmabuf.c

```diff
@@ -0,0 +1,362 @@
+ */
+
+#define _POSIX_C_SOURCE 200112L
+#include <errno.h>
+#include <fcntl.h>
+#include <limits.h>
+#include <png.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/mman.h>
+#include <sys/param.h>
+#include <sys/stat.h>
+#include <sys/wait.h>
+#include <unistd.h>
+#include <assert.h>
+#include <gbm.h>
+#include <xf86drm.h>
+#include <drm_fourcc.h>
+#include <wayland-client-protocol.h>
+#include "wlr-screencopy-unstable-v1-client-protocol.h"
+#include "linux-dmabuf-unstable-v1-client-protocol.h"
+
+struct format {
+	uint32_t format;
+	bool is_bgr;
+};
+
+static int drm_fd = -1;
+static struct gbm_device *gbm_device = NULL;
+
+static struct zwp_linux_dmabuf_v1 *dmabuf = NULL;
+static struct zwlr_screencopy_manager_v1 *screencopy_manager = NULL;
+static struct wl_output *output = NULL;
+
+static struct {
+	struct gbm_bo *bo;
+	struct wl_buffer *wl_buffer;
+	int width, height;
+	uint32_t format;
+	bool y_invert;
+} buffer;
+
+bool buffer_copy_done = false;
+static bool have_linux_dmabuf = false;
+
+// wl_shm_format describes little-endian formats, libpng uses big-endian
+// formats (so Wayland's ABGR is libpng's RGBA).
+static const struct format formats[] = {
+	{DRM_FORMAT_XRGB8888, true},
+	{DRM_FORMAT_ARGB8888, true},
+	{DRM_FORMAT_XBGR8888, false},
+	{DRM_FORMAT_ABGR8888, false},
+};
+
+static bool find_render_node(char *node, size_t maxlen) {
+	bool r = false;
+	drmDevice *devices[64];
+
+	int n = drmGetDevices2(0, devices, sizeof(devices) / sizeof(devices[0]));
+	for (int i = 0; i < n; ++i) {
+		drmDevice *dev = devices[i];
+		if (!(dev->available_nodes & (1 << DRM_NODE_RENDER))) {
+			continue;
+		}
+
+		strncpy(node, dev->nodes[DRM_NODE_RENDER], maxlen);
+		node[maxlen - 1] = '\0';
+		r = true;
+		break;
+	}
+
+	drmFreeDevices(devices, n);
+	return r;
+}
+
+static void dmabuf_created(void *data,
+		struct zwp_linux_buffer_params_v1 *params,
+		struct wl_buffer *wl_buffer) {
+	buffer.wl_buffer = wl_buffer;
+
+	zwlr_screencopy_frame_v1_copy(data, buffer.wl_buffer);
+}
+
+static void dmabuf_failed(void *data,
+		struct zwp_linux_buffer_params_v1 *params) {
+	fprintf(stderr, "Failed to create dmabuf\n");
+	exit(EXIT_FAILURE);
+}
+
+static const struct zwp_linux_buffer_params_v1_listener params_listener = {
+	.created = dmabuf_created,
+	.failed = dmabuf_failed,
+};
+
+static void frame_handle_buffer(void *data,
+		struct zwlr_screencopy_frame_v1 *frame, uint32_t wl_format,
+		uint32_t width, uint32_t height, uint32_t stride) {
+	// Not implemented
+}
+
+static void frame_handle_linux_dmabuf(void *data,
+		struct zwlr_screencopy_frame_v1 *frame, uint32_t fourcc,
+		uint32_t width, uint32_t height) {
+	buffer.width = width;
+	buffer.height = height;
+	buffer.format = fourcc;
+	have_linux_dmabuf = true;
+}
+
+static void frame_handle_buffer_done(void *data,
+		struct zwlr_screencopy_frame_v1 *frame) {
+	assert(!buffer.bo);
+
+	if (!have_linux_dmabuf) {
+		fprintf(stderr, "\n");
```

Comment:
Maybe print a useful error message here?

Reply:
Hah! Forgot to fill in that one.